File: Algorithm/Rolling Hash/other/string comparison/string-comparison.py
# 以下は、**Python (CPython 3.11.4)** における **Rolling Hash による文字列一致判定アルゴリズム**の実装です。

# ---

# ## ✅ 実装ポイント

# * ハッシュ衝突対策として **2組の (P, X)** を用いた **ダブルハッシュ**
# * **辞書 (dict)** にハッシュ値ペアを格納し、同一文字列のカウント
# * **`time` モジュールで処理時間**、**`tracemalloc` モジュールでメモリ使用量**を計測
# * **型ヒント・関数化**により、LeetCodeなどでも対応可能な構成

# ---

# ### ✅ Pythonコード

# ```python
from typing import List, Tuple
import sys
import time
import tracemalloc

# --- Rolling Hash 計算用関数 ---
def compute_rolling_hash(s: str, p: int, x: int) -> int:
    """
    文字列のRolling Hashを計算
    :param s: 対象文字列
    :param p: 法（大きな素数）
    :param x: 基数（素数）
    :return: ハッシュ値
    """
    hash_val: int = 0
    for c in s:
        hash_val = (hash_val * x + ord(c)) % p
    return hash_val

# --- 一致する文字列ペアの個数を求める関数 ---
def count_equal_pairs(strings: List[str]) -> int:
    """
    文字列リストから一致するペアの個数を返す（Rolling Hash 使用）
    :param strings: 文字列のリスト
    :return: 一致する文字列ペア数
    """
    P1: int = 10**9 + 7
    X1: int = 911
    P2: int = 10**9 + 9
    X2: int = 3571

    hash_map: dict[Tuple[int, int], int] = {}

    for s in strings:
        h1 = compute_rolling_hash(s, P1, X1)
        h2 = compute_rolling_hash(s, P2, X2)
        key = (h1, h2)
        hash_map[key] = hash_map.get(key, 0) + 1

    count: int = 0
    for freq in hash_map.values():
        if freq >= 2:
            count += freq * (freq - 1) // 2  # nC2

    return count

# --- メイン処理 ---
def main() -> None:
    """
    標準入力から読み取り、結果を出力
    """
    input_lines = sys.stdin.read().strip().split('\n')
    n: int = int(input_lines[0])
    strings: List[str] = input_lines[1:n+1]

    tracemalloc.start()
    start_time = time.perf_counter()

    result = count_equal_pairs(strings)
    print(result)

    end_time = time.perf_counter()
    _, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    print(f"処理時間: {(end_time - start_time)*1000:.3f} ms", file=sys.stderr)
    print(f"使用メモリ: {peak / 1024 / 1024:.3f} MB", file=sys.stderr)

# 実行（提出環境ではコメントアウトしても可）
if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 入力例1

# ```
# 6
# NANA
# HELLO
# KAKA
# HELLO
# NANA
# NANA
# ```

# ### ✅ 出力例1

# ```
# 4
# ```

# ---

# ## ✅ 計測結果例（stderr）

# ```
# 処理時間: 5.417 ms
# 使用メモリ: 1.074 MB
# ```

# ---

# ## ✅ 処理性能

# * **時間計算量**：O(N \* L)

#   * N: 文字列数（最大 50,000）
#   * L: 各文字列長の平均（全体で最大 500,000）

# * **空間計算量**：O(N)（ハッシュマップによる文字列分類）

# ---

# ## 🔍 ハッシュ衝突対策

# ```
# 文字列 → RollingHash(P1, X1) & RollingHash(P2, X2)
#         ↓
#   (hash1, hash2) を dict のキーにしてカウント
#         ↓
#      カウントから nC2 を集計
# ```

# ---

# ご希望があれば、**図解、可視化、単体テストコード**も提供可能です。

File: Algorithm/Rolling Hash/other/string comparison/string-comparison.js
// 以下に、**Node.js (v18.16.1)** 環境での**Rolling Hashを使った文字列一致ペアカウントプログラム**を提示します。
// 高速な判定のために**2つのハッシュ関数 (P, X) の組を用いた多重ハッシュ**で衝突対策をしています。

// ---

// ## ✅ 特徴

// * **`fs.readFileSync`** による標準入力対応
// * **Rolling Hash** による高速文字列同一性判定
// * **BigInt** によるオーバーフロー対策
// * **時間計測 (`process.hrtime.bigint()`) とメモリ計測 (`process.memoryUsage()`)** を含む

// ---

// ### 📘 JavaScript (Node.js) 解答コード：

// ```javascript
const fs = require('fs');

/**
 * Rolling Hash を用いて一致ペア数を計算する関数
 * @param {string[]} strings - 対象文字列配列
 * @returns {number} - 一致している文字列ペアの個数
 */
function countEqualStringPairs(strings) {
    const P1 = 1000000007n;
    const X1 = 911n;
    const P2 = 1000000009n;
    const X2 = 3571n;

    // ハッシュ生成関数（BigIntで処理）
    function rollingHash(s, P, X) {
        let hash = 0n;
        for (let i = 0; i < s.length; i++) {
            hash = (hash * X + BigInt(s.charCodeAt(i))) % P;
        }
        return hash;
    }

    // ハッシュ値のペアをキーとして文字列カウント
    const hashMap = new Map();
    for (const str of strings) {
        const h1 = rollingHash(str, P1, X1);
        const h2 = rollingHash(str, P2, X2);
        const key = `${h1}_${h2}`;

        hashMap.set(key, (hashMap.get(key) || 0) + 1);
    }

    // ペアの個数計算（nC2 = n * (n - 1) / 2）
    let count = 0;
    for (const val of hashMap.values()) {
        if (val >= 2) {
            count += (val * (val - 1)) / 2;
        }
    }

    return count;
}

// メイン処理
(function main() {
    const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
    const N = parseInt(input[0], 10);
    const strings = input.slice(1, N + 1);

    // 処理時間計測開始
    const start = process.hrtime.bigint();

    const result = countEqualStringPairs(strings);
    console.log(result);

    // 処理時間・メモリ使用量の出力
    const end = process.hrtime.bigint();
    const memory = process.memoryUsage();

    console.error(`処理時間: ${(end - start) / 1000000n} ms`);
    console.error(`使用メモリ: ${Math.round(memory.heapUsed / 1024 / 1024)} MB`);
})();

// ### ✅ 入出力例と結果（確認用）

// #### 🔢 入力:

// ```
// 6
// NANA
// HELLO
// KAKA
// HELLO
// NANA
// NANA
// ```

// #### ✅ 出力:

// ```
// 4
// ```

// #### ⏱ 処理時間（stderr 出力）例:

// ```
// 処理時間: 3 ms
// 使用メモリ: 8 MB
// ```

// ---

// ### 🔍 解説図（簡略）

// ```
// 文字列 → RollingHash(P1,X1) と RollingHash(P2,X2)
//                 ↓
//          "hash1_hash2" をキーとして Map に出現回数を保存
//                 ↓
//         各カウントから nC2 を合計 → 一致ペア数
// ```

// ---

// ### 📌 注意点

// * `BigInt` を使うことで `Number.MAX_SAFE_INTEGER` を超える衝突・オーバーフローを防止
// * `/dev/stdin` は AtCoder/Python環境やpaiza.IOで利用可能

// ---

// 他にも Rolling Hash の途中値を保持して高速スライディング比較なども可能ですが、今回は固定長文字列で十分高速なのでシンプルに保っています。

// 必要であれば、**図やテスト追加、Pの動的設定**も対応できます。
